core/src/impl/Kokkos_EBO.hpp

```diff
@@ -61,37 +61,73 @@
 namespace Kokkos {
 namespace Impl {
 
-template <class T, bool>
-struct EBOBaseImpl;
+template <int I>
+struct NotOnDeviceCtorDisambiguator { };
+
+template <class... Args>
+struct NoCtorsNotOnDevice : std::false_type { };
+
+template <class... Args>
+struct DefaultCtorNotOnDevice : std::false_type { };
+
+template <>
+struct DefaultCtorNotOnDevice<> : std::true_type { };
 
-template <class T>
-struct EBOBaseImpl<T, true> {
+template <class T, bool Empty, template <class...> class CtorNotOnDevice = NoCtorsNotOnDevice>
+struct EBOBaseImpl;
 
+template <class T, template <class...> class CtorNotOnDevice>
+struct EBOBaseImpl<T, true, CtorNotOnDevice> {
+
+  /*
+   * Workaround for constexpr in C++11: we need to still call T(args...), but we
+   * can't do so in the body of a constexpr function (in C++11), and there's no
+   * data member to construct into. But we can construct into an argument
+   * of a delegating constructor...
+   */
+  // TODO @minor DSH the destructor gets called too early with this workaround
+  struct _constexpr_14_workaround_tag { };
+  struct _constexpr_14_workaround_no_device_tag { };
   KOKKOS_FORCEINLINE_FUNCTION
-  constexpr EBOBaseImpl() noexcept = default;
+  constexpr EBOBaseImpl(_constexpr_14_workaround_tag, T&&) noexcept { }
+  inline constexpr EBOBaseImpl(_constexpr_14_workaround_no_device_tag, T&&) noexcept { }
```

Comment:
To avoid the having this function count against the compiler's "non-`inline`d" inlining budget.  The compiler's `constexpr` budget for inlining is a separate thing, so it's useful to put both.